### PR TITLE
Reader: basics for recommendations/posts stream

### PR DESCRIPTION
**Pull request by aaronyan**
A basic page to view streams of recommended posts for testing the Data team's Cold-start and Warm-start algorithms. These algorithms are used to populate the Reader with posts that take into account a user's interactions with WordPress.com.

- Cold-start algorithms = recommend posts for new users (using no information about user interactions (likes/comments/posts) with WordPress.com)
- Warm-start algorithms = recommend posts using user interactions (likes/comments/posts) with WordPress.com

### CODE REVIEW

File: client/reader/index.js [hunk 1]
```diff
@@ -63,7 +63,14 @@ module.exports = function() {
 		page( '/tag/:tag', updateLastRoute, controller.removePost, controller.sidebar, controller.tagListing );
 	}
 
-	page( '/read/a8c', updateLastRoute, controller.removePost, controller.sidebar, forceTeamA8C, controller.readA8C );
+	if ( config.isEnabled( 'reader/a8c' ) ) {
+		page( '/read/a8c', updateLastRoute, controller.removePost, controller.sidebar, forceTeamA8C, controller.readA8C );
+	}
+
+	if ( config.isEnabled( 'reader/warmstart' ) ) {
```
Comment (mtias) [inline client/reader/index.js]: It'd be nice to start this as a separate section.

Comment (aaronyan) [inline client/reader/index.js]: Thanks @mtias! I think separating this out is a great idea. Done and done!

Separated out the feature check of /read/a8c: https://github.com/Automattic/wp-calypso/pull/5188

**Comment by aaronyan** [comment]
Ready for review @blowery! Thank you for the help during this PR!

Before merging the PR, we will need the /read/recommendations/warm-start endpoint to be deployed: https://code.a8c.com/D1687

I will comment here when the warm-start endpoint is up.

File: client/reader/recommendations/controller.js [hunk 2]
```diff
@@ -10,10 +10,16 @@ import ReactDom from 'react-dom';
...
 import titleActions from 'lib/screen-title/actions';
 import i18n from 'lib/mixins/i18n';
-import { trackPageLoad } from 'reader/controller-helper';
+import { ensureStoreLoading, trackPageLoad, trackUpdatesLoaded, setPageTitle } from 'reader/controller-helper';
+import route from 'lib/route';
+import feedStreamFactory from 'lib/feed-stream-store';
 
 const ANALYTICS_PAGE_TITLE = 'Reader';
 
+function userHasHistory( context ) {
```
Comment (blowery) [inline client/reader/recommendations/controller.js]: maybe move this to reader/controller-helper

File: client/reader/recommendations/controller.js [hunk 3]
```diff
@@ -36,5 +42,36 @@ export default {
 
 		trackPageLoad( basePath, fullAnalyticsPageTitle, mcKey );
 		titleActions.setTitle( i18n.translate( 'Recommended Sites For You ‹ Reader' ) );
+	},
+
+	// Post Recommendations - Used by the Data team to test recommendation algorithms
+	recommendedPosts( context ) {
+		var WarmstartStream = require( 'reader/recommendations/posts' ),
+			basePath = route.sectionify( context.path ),
+			fullAnalyticsPageTitle = ANALYTICS_PAGE_TITLE + ' > Warmstart',
```
Comment (blowery) [inline client/reader/recommendations/controller.js]: warm start is an internal name, i don't think we really want it in stats or variable names

File: client/lib/feed-stream-store/index.js [hunk 4]
```diff
@@ -140,6 +140,19 @@ function getStoreForFeatured( storeId ) {
 	} );
 }
 
+function getStoreForWarmstart( storeId ) {
```
Comment (blowery) [inline client/lib/feed-stream-store/index.js]: see other comment, but warmStart is an internal name. we should probably call this what is is, recommended posts.

File: client/sections.js [hunk 5]
```diff
@@ -222,6 +222,14 @@ if ( config.isEnabled( 'reader' ) ) {
 	} );
 
 	sections.push( {
+		name: 'reader-post-recomendations',
+		paths: [ '/recommendations' ],
```
Comment (blowery) [inline client/sections.js]: i think this needs to be more specific, and possibly above the other recs section. What gets loaded when you try to load recs?

File: client/lib/feed-stream-store/index.js [hunk 6]
```diff
@@ -170,6 +183,8 @@ function feedStoreFactory( storeId ) {
 			onUpdateFetch: limitSiteParamsForLikes,
 			dateProperty: 'date_liked'
 		} );
+	} else if ( storeId.indexOf( 'warmstart' ) === 0 ) {
```
Comment (blowery) [inline client/lib/feed-stream-store/index.js]: should just be a straight equivalence if there's no parameter after a :

File: client/sections.js (hunk 5 above)
Comment (aaronyan) [inline client/sections.js]: Good call. I've made this more specific on the path (/recommendations/posts) and placed above the other recs. The original Recommendations gets loaded correctly in both scenarios (when /posts is above and below the original /recommendations).

**Comment by aaronyan** [comment]
Thanks for the great comments @blowery. Added all of your suggested changes!

**Comment by aaronyan** [comment]
The /read/recommendations/warm-start endpoint has been deployed! D1687-code. @blowery, the PR ready for merge?

File: client/lib/feed-stream-store/index.js [hunk 7]
```diff
@@ -140,6 +140,19 @@ function getStoreForFeatured( storeId ) {
...
+function getStoreForRecommendedPosts( storeId ) {
+	var fetcher = function( query, callback ) {
+			wpcomUndoc.readRecommendedPosts( query, callback );
+		};
+
+	return new PagedStream( {
+		id: storeId,
+		fetcher: fetcher,
+		keyMaker: siteKeyMaker,
+		perPage: 5
```
Comment (blowery) [inline client/lib/feed-stream-store/index.js]: we generally use 7, but 5 is probably fine too

**Comment by blowery** [comment]
Trying this out 👀

**Comment by blowery** [comment]
This is amazing.

File: client/reader/recommendations/posts/index.jsx [hunk 8]
```diff
@@ -0,0 +1,26 @@
+var React = require( 'react' );
```
Comment (mtias) [inline client/reader/recommendations/posts/index.jsx]: This file should probably use ES6. Also the namespace is not quite right. It should be `posts-recommendations` or similar but not just `posts`.

**Comment by mtias** [comment]
@blowery @aaronyan just left a comment on the `recommendations/posts` component. It'd be nice to fix that namespace.

File: client/reader/recommendations/posts/index.jsx (hunk 8 above)
Comment (blowery) [inline client/reader/recommendations/posts/index.jsx]: namespace?

Comment (mtias) [inline client/reader/recommendations/posts/index.jsx]: https://wpcalypso.wordpress.com/devdocs/docs/components.md

> Classes (both in components and sub-components) are always prefixed to the component name.

Which comes directly from the folder name.

Comment (mtias) [inline client/reader/recommendations/posts/index.jsx]: If the folder is `posts` then the classes would be `posts__`, which shows the namespace is too generic and clashes with other "posts" components, so the folder name should be updated to `posts-recommendations` so we have `posts-recommendations__` as the class prefix.

Comment (blowery) [inline client/reader/recommendations/posts/index.jsx]: so `reader/recommendations/posts-recommendations`? that seems redundant. maybe we should update the naming guidelines to take subfolders into account?

Comment (mtias) [inline client/reader/recommendations/posts/index.jsx]: The main point about the guidelines is that sub-folders are not taken into account, and any component is considered to coexist under a flat hierarchy. (That gives us the flexibility that any folder is completely autonomous and we can group them as needed.) In this case I'd just place `posts-recommendations` in `/reader/`'s root or make posts-recommendations a sub-component of `reader/recommendations` if it's not meant to be used elsewhere at all.

Not caring about folder structure makes it really straightforward how a class should be named — it only has to look at its folder name.

Comment (aaronyan) [inline client/reader/recommendations/posts/index.jsx]: Thanks @mtias! I can look into adjusting the folders to match the convention. @blowery, do we have a preference on one vs the other? I like the /reader/recommendations/posts URL so i'm leaning towards the posts.jsx route.

**Pull request merged**